lib/client/src/context.rs

```diff
@@ -554,6 +628,92 @@ impl MangoGroupContext {
         let new_perp_markets = fetch_perp_markets(rpc, mango_v4::id(), self.group).await?;
         Ok(new_perp_markets.len() > self.perp_markets.len())
     }
+
+    /// Returns a map of oracle pubkey -> FallbackOracleContext
+    pub async fn derive_fallback_oracle_keys(
+        &self,
+        fallback_oracle_config: &FallbackOracleConfig,
+        rpc: &RpcClientAsync,
+    ) -> anyhow::Result<HashMap<Pubkey, FallbackOracleContext>> {
+        // FUTURE: implement for perp oracles as well
+        let fallbacks_by_oracle = match fallback_oracle_config {
+            FallbackOracleConfig::Never => HashMap::new(),
+            FallbackOracleConfig::Fixed(keys) => self
+                .tokens
+                .iter()
+                .filter(|token| {
+                    token.1.fallback_context.key != Pubkey::default()
+                        && keys.contains(&token.1.fallback_context.key)
+                })
+                .map(|t| (t.1.oracle, t.1.fallback_context.clone()))
+                .collect(),
+            FallbackOracleConfig::All => self
+                .tokens
+                .iter()
+                .filter(|token| token.1.fallback_context.key != Pubkey::default())
+                .map(|t| (t.1.oracle, t.1.fallback_context.clone()))
+                .collect(),
+            FallbackOracleConfig::Dynamic => {
+                let tokens_by_oracle: HashMap<Pubkey, &TokenContext> =
+                    self.tokens.iter().map(|t| (t.1.oracle, t.1)).collect();
+                let oracle_keys: Vec<Pubkey> =
+                    tokens_by_oracle.values().map(|b| b.oracle).collect();
+                let oracle_accounts = fetch_multiple_accounts(rpc, &oracle_keys).await?;
+                let now_slot = rpc.get_slot().await?;
+
+                let mut stale_oracles_with_fallbacks = vec![];
+                for acc in oracle_accounts {
+                    let token = tokens_by_oracle.get(acc.key()).unwrap();
+                    let state = oracle_state_unchecked(
+                        &OracleAccountInfos::from_reader(&acc),
+                        token.decimals,
+                    )?;
+                    let oracle_is_valid = state
+                        .check_confidence_and_maybe_staleness(&token.oracle_config, Some(now_slot));
+                    if oracle_is_valid.is_err() && token.fallback_context.key != Pubkey::default() {
+                        stale_oracles_with_fallbacks
+                            .push((token.oracle, token.fallback_context.clone()));
+                    }
+                }
+                stale_oracles_with_fallbacks.into_iter().collect()
+            }
+        };
+
+        Ok(fallbacks_by_oracle)
+    }
+
+    /// Returns the a map of oracle pubkey -> (fallback_oracle_key, fallback_oracle_quote_key).
+    pub fn derive_fallback_oracle_keys_sync(
```

Comment:
I hate to have an async and sync version, but the async version was causing thread safety issues in the settler. Specifically this thread: https://github.com/blockworks-foundation/mango-v4/blob/53bc52efe7c0b3eee01d4737be1f043c43cf9980/bin/settler/src/main.rs#L316 would throw `rust could not prove `[async block@bin/settler/src/main.rs:320:9: 335:10]: std::marker::Send`

Making `token_withdraw_instructions` async here: https://github.com/blockworks-foundation/mango-v4/blob/53bc52efe7c0b3eee01d4737be1f043c43cf9980/bin/settler/src/tcs_start.rs#L118 is what causes the issue, but I am still unable to understand why that's the case? Ultimatley I think something about the async version of derive_fallback_oracle_keys should be to blame but I can't pinpoint it.

Here's a similar Rust issue for more context: https://github.com/rust-lang/rust/issues/102211

Reply:
Yeah, those are a nightmare to debug. Basically what I did was to go through the callchain of the presumed culprit and checking if any of the context of the calls might not be Sent. After a bunch of staring and a few false leads, I noticed that the `.unique()` iterator function might very well return a non-send wrapper and made that into a vec to make it Send.

Reply:
A helpful thing I had noticed early was that derive_fallback_oracle_keys() wasn't the problem, it was token_withdraw_instructions() being async that caused the issue.